Kuadrant/kuadrant-operator
# Istio ext authz service config

### what

Authorino set up in Istio as a Auth GRPC service 

Fixes https://github.com/Kuadrant/kuadrant-controller/issues/149

* add Gateway API CRDs dependency
* istio install makefile targets using the operator
* Upgrade installed Istio to 1.13.2. Isito 1.12.X has an [issue](https://github.com/istio/istio/issues/36114).
* Implemented in the controller:
  * setup logger: support for log levels (`debug` →  `info` →  `error`) and configurable log modes: `production`, `development`
  * status conditions: `Ready` condition
  * reconcile Istio mesh configuration for the kuadrant external authorizer. When the kuadrant CR is deleted, the configuration is cleaned up. Currently, the external authorizer address is hardcoded to 
```yaml
- envoyExtAuthzGrpc:
    port: 50051
    service: authorino-authorino-authorization.kuadrant-system.svc.cluster.local
  name: kuadrant-authorization
```

  * The name/NS of the IstioOperator CR used to configure external authorizer is read from en var

|   | Environemnt Variable Name | Default Value |
| --- | --- | --- | 
|  IstioOperator name | `ISTIOOPERATOR_NAME` | `istiocontrolplane` |
|  IstioOperator namespace | `ISTIOOPERATOR_NAMESPACE` | `istio-system` |

### verification steps

create kuadrant cluster locally with kind

```
make kind-create-kuadrant-cluster
```

Check istio config does not have any kuadrant's external authorizer. For that, the `istiocontrolplane` named IstioOperator CR is checked:
```
k get istiooperator istiocontrolplane -n istio-system -o jsonpath='{.spec.meshConfig.extensionProviders}'
```
The result should be empty.

Create empty Kuadrant CR object

```yaml
kubectl apply -f - <<EOF
---
apiVersion: kuadrant.kuadrant.io/v1beta1
kind: Kuadrant
metadata:
  name: kuadrant-sample
spec: {}
EOF
```

Check istio config **does** have kuadrant¡s external authorizer. For that, the `istiocontrolplane` named IstioOperator CR is checked:
```yaml
k get istiooperator istiocontrolplane -n istio-system -o jsonpath='{.spec.meshConfig.extensionProviders}'
- envoyExtAuthzGrpc:
    port: 50051
    service: authorino-authorino-authorization.kuadrant-system.svc.cluster.local
  name: kuadrant-authorization
```

Delete the kuadrant CR
```
k delete kuadrant kuadrant-sample
```
Check istio config does no longer have any kuadrant's external authorizer. For that, the `istiocontrolplane` named IstioOperator CR is checked:
```
k get istiooperator istiocontrolplane -n istio-system -o jsonpath='{.spec.meshConfig.extensionProviders}'
[]
```
The result should be empty (or empty array)
